common/core/nesteddisplay.cpp

```diff
@@ -67,10 +133,109 @@ bool NestedDisplay::SetPowerMode(uint32_t /*power_mode*/) {
   return true;
 }
 
-bool NestedDisplay::Present(std::vector<HwcLayer *> & /*source_layers*/,
+bool NestedDisplay::Present(std::vector<HwcLayer *> &source_layers,
                             int32_t * /*retire_fence*/,
                             bool /*handle_constraints*/) {
-  // TODO implement Present
+
+  int ret = 0;
+  size_t size = source_layers.size();
+  const uint32_t* pitches;
+  const uint32_t* offsets;
+  HWCNativeHandle sf_handle;
+  size_t buffer_number = 0;
+  size_t info_size = sizeof(vm_buffer_info);
+  size_t header_size = sizeof(vm_header);
+
+  for (size_t layer_index = 0; layer_index < size; layer_index++) {
+    HwcLayer* layer = source_layers.at(layer_index);
+    if (!layer->IsVisible())
+      continue;
+
+    HwcRect<int> display_frame = layer->GetDisplayFrame();
+    sf_handle = layer->GetNativeHandle();
+    auto search = mHyperDmaExportedBuffers.find(sf_handle);
```

Comment:
You want to use caching like overlaylayer does here:
https://github.com/intel/IA-Hardware-Composer/blob/master/common/core/overlaylayer.cpp#L72